Make WebGL rendering context creation fail if GPU is not available

pvollan opened the pull request:
#### 6ed0df1df12a2ae89b0937a49030f6e97ec1b2d4
<pre>
Make WebGL rendering context creation fail if GPU is not available
<a href="https://bugs.webkit.org/show_bug.cgi?id=242814">https://bugs.webkit.org/show_bug.cgi?id=242814</a>
&lt;rdar://96096307&gt;

Reviewed by Geoffrey Garen.

We are seeing WebGL crash reports where no Metal devices appear to be available. Guard against
this crash by failing to create WebGL contexts when Metal is not available.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContextWebGL):
* Source/WebCore/platform/graphics/gpu/GPUAvailability.h: Added.
* Source/WebCore/platform/graphics/gpu/cocoa/GPUAvailabilityMetal.mm: Added.
(WebCore::isGPUAvailable):

Canonical link: <a href="https://commits.webkit.org/252531@main">https://commits.webkit.org/252531@main</a>
</pre>

cdumez commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -453,6 +455,13 @@ WebGLRenderingContextBase* HTMLCanvasElement::createContextWebGL(WebGLVersion ty
         return nullptr;
 
+#if PLATFORM(MAC)
+    if (!isGPUAvailable()) {
```
Why is it relevant for the WebProcess to have Metal access when we do WebGL in the GPUProcess nowadays?

geoffreygaren replied:
WebGL runs in the WebContent process on macOS.

cdumez replied:
Then this should be probably controlled based on some sort of feature flag, not a PLATFORM(MAC) which will bite us later when we do add support on macOS.

cdumez replied:
Like add a WebCore binding for WebKit's UseGPUProcessForWebGLEnabled feature flag and rely on this instead of a `#if PLATFORM(MAC)` compile flag.

geoffreygaren replied:
There's a setting: UseGPUProcessForWebGLEnabled. Let's use that.

pvollan replied:
> Like add a WebCore binding for WebKit's UseGPUProcessForWebGLEnabled feature flag and rely on this instead of a `#if PLATFORM(MAC)` compile flag.

Yes, that is a good point.

Thanks for reviewing!

pvollan replied:
> There's a setting: UseGPUProcessForWebGLEnabled. Let's use that.

Fixed in latest patch.

Thanks for reviewing!

geoffreygaren reviewed (approved):
r=me

Please change to checking the UseGPUProcessForWebGLEnabled setting rather than PLATFORM(MAC).

cdumez commented on Source/WebCore/platform/graphics/gpu/cocoa/GPUAvailabilityMetal.mm:
```diff
@@ -0,0 +1,48 @@
+
+bool isGPUAvailable()
+{
+    static dispatch_once_t once = 0;
```
`= 0` should not be needed.

pvollan replied:
> `= 0` should not be needed.

Thanks, fixed!

pvollan committed:
Make WebGL rendering context creation fail if GPU is not available
https://bugs.webkit.org/show_bug.cgi?id=242814
<rdar://96096307>

Reviewed by Geoffrey Garen.

We are seeing WebGL crash reports where no Metal devices appear to be available. Guard against
this crash by failing to create WebGL contexts when Metal is not available.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContextWebGL):
* Source/WebCore/platform/graphics/gpu/GPUAvailability.h: Added.
* Source/WebCore/platform/graphics/gpu/cocoa/GPUAvailabilityMetal.mm: Added.
(WebCore::isGPUAvailable):

Canonical link: https://commits.webkit.org/252531@main

shallawa commented on Source/WebCore/platform/graphics/gpu/cocoa/GPUAvailabilityMetal.mm:
```diff
@@ -0,0 +1,48 @@
+
+namespace WebCore {
+
+bool isGPUAvailable()
+{
+    static dispatch_once_t once;
+    static bool metalAvailable = false;
+    dispatch_once(&once, ^{
+        auto devices = adoptNS(MTLCopyAllDevices());
+        metalAvailable = devices && [devices.get() count] > 0;
+    });
+    return metalAvailable;
+}
```
Should not this function be guarded by #if HAVE(GPU_AVAILABILITY_CHECK)? I think MTLCopyAllDevices() is not available on iOS.

pvollan replied:
> Should not this function be guarded by #if HAVE(GPU_AVAILABILITY_CHECK)? I think MTLCopyAllDevices() is not available on iOS.

Yes, that is addressed in https://github.com/WebKit/WebKit/pull/2502.

Thanks for reviewing!
